File: src/main/java/com/github/vendigo/acetest/assertion/ExceptionMatcher.java
package com.github.vendigo.acetest.assertion;

import java.io.PrintWriter;
import java.io.StringWriter;
import java.util.Optional;
import java.util.stream.Stream;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.containsString;
import static org.hamcrest.Matchers.notNullValue;

public class ExceptionMatcher {
    private static final String LINE_SEPARATOR = "\n";
    private static final String CAUSED_BY = "Caused by: ";

    public static void matchException(Throwable throwable, String stackTraceContains) {
        assertThat(throwable, notNullValue());
        StringWriter stringWriter = new StringWriter();
        throwable.printStackTrace(new PrintWriter(stringWriter));
        String stackTrace = stringWriter.toString();
        Optional<String> exceptionString = Stream.of(stackTrace.split(LINE_SEPARATOR))
                .filter(s -> s.contains(CAUSED_BY))
                .findFirst();
        if (exceptionString.isPresent()) {
            assertThat(exceptionString.get(), containsString(stackTraceContains));
        } else {
            throw new AssertionError("Stacktrace doesn't contain 'Caused by section'");
        }
    }
}

File: src/main/java/com/github/vendigo/acetest/assertion/ExceptionMatcher.kt
package com.github.vendigo.acetest.assertion

import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.containsString
import org.hamcrest.Matchers.notNullValue
import java.io.PrintWriter
import java.io.StringWriter

private val LINE_SEPARATOR = "\n"
private val CAUSED_BY = "Caused by: "

fun matchException(throwable: Throwable?, stackTraceContains: String) {
    assertThat(throwable, notNullValue())
    val stringWriter = StringWriter()
    throwable!!.printStackTrace(PrintWriter(stringWriter))
    val stackTrace = stringWriter.toString()
    val exceptionString = stackTrace.split(LINE_SEPARATOR.toRegex())
            .firstOrNull { s -> s.contains(CAUSED_BY) }
    if (exceptionString != null) {
        assertThat(exceptionString, containsString(stackTraceContains))
    } else {
        throw AssertionError("Stacktrace doesn't contain 'Caused by' section")
    }
}

